src/vs/workbench/browser/parts/editor/multiRowEditorTabsControl.ts

```diff
@@ -0,0 +1,172 @@
+import { IEditorTitleControlDimensions } from 'vs/workbench/browser/parts/editor/editorTitleControl';
+import { IReadonlyEditorGroupModel } from 'vs/workbench/common/editor/editorGroupModel';
+
+export class MultiRowEditorControl extends Disposable implements IEditorTabsControl {
+
+	private readonly stickyEditorTabsControl: IEditorTabsControl;
+	private readonly unstickyEditorTabsControl: IEditorTabsControl;
+
+	constructor(
+		private parent: HTMLElement,
+		private accessor: IEditorGroupsAccessor,
+		private group: IEditorGroupView,
+		private model: IReadonlyEditorGroupModel,
+		@IInstantiationService protected instantiationService: IInstantiationService
+	) {
+		super();
+
+		this.stickyEditorTabsControl = this._register(this.instantiationService.createInstance(MultiEditorTabsControl, this.parent, this.accessor, this.group, new StickyEditorGroupModel(this.model)));
+		this.unstickyEditorTabsControl = this._register(this.instantiationService.createInstance(MultiEditorTabsControl, this.parent, this.accessor, this.group, new UnstickyEditorGroupModel(this.model)));
+
+		this.handlePinnedTabsSeparateRowToolbars();
+	}
+
+	private handlePinnedTabsSeparateRowToolbars(): void {
+		if (this.group.count === 0) {
+			// Do nothing as no tab bar is visible
+			return;
+		}
+		// Ensure action toolbar is only visible once
+		if (this.group.count === this.group.stickyCount) {
+			this.parent.classList.toggle('second-tabbar-visible', false);
+		} else {
+			this.parent.classList.toggle('second-tabbar-visible', true);
+		}
+	}
+
+	private getEditorTabsController(editor: EditorInput): IEditorTabsControl {
+		return this.model.isSticky(editor) ? this.stickyEditorTabsControl : this.unstickyEditorTabsControl;
+	}
+
+	openEditor(editor: EditorInput): boolean {
+		const [editorTabController, otherTabController] = this.model.isSticky(editor) ? [this.stickyEditorTabsControl, this.unstickyEditorTabsControl] : [this.unstickyEditorTabsControl, this.stickyEditorTabsControl];
+		const didChange = editorTabController.openEditor(editor);
+		if (didChange) {
+			// HACK: To render all editor tabs on startup, otherwise only one row gets rendered
+			otherTabController.openEditors([]);
+		}
+		return didChange;
+	}
+
+	openEditors(editors: EditorInput[]): boolean {
+		const stickyEditors = editors.filter(e => this.model.isSticky(e));
+		const unstickyEditors = editors.filter(e => !this.model.isSticky(e));
+
+		const didChangeOpenEditorsSticky = this.stickyEditorTabsControl.openEditors(stickyEditors);
+		const didChangeOpenEditorsUnSticky = this.unstickyEditorTabsControl.openEditors(unstickyEditors);
+
+		return didChangeOpenEditorsSticky || didChangeOpenEditorsUnSticky;
+	}
+
+	beforeCloseEditor(editor: EditorInput): void {
+		this.getEditorTabsController(editor).beforeCloseEditor(editor);
+	}
+
+	closeEditor(editor: EditorInput): void {
+		// Has to be called on both tab bars
```

Comment:
Why on both?

Reply:
As the editor does not store any information on it's index and sticky state and the model has removed the state of the closed editor already, we can't figure out if the closed editor was a sticky editor or not